src/opencl_kde_depth_packet_processor.cl

```diff
@@ -0,0 +1,769 @@
+      ind_third = ind_second;
+      err_min_second = err;
+      ind_second = i;
+    }
+    else if(err<err_min_third)
+    {
+      err_min_third = err;
+      ind_third = i;
+    } 
+  }
+
+  //decode ind_min
+  float mvals = m_list[ind_min];
+  float nvals = n_list[ind_min];
+  float kvals = k_list[ind_min];
+
+  //Weighted phases for phase fusion weighted average
+  float phi2_out = (t2 / 2.0f + mvals);
+  float phi1_out = (t1 / 15.0f + kvals);
+  float phi0_out = (t0 / 3.0f + nvals);
+
+  *err_w1 = err_min;
+
+  *phase_first = (phi2_out + phi1_out + phi0_out) / 3.0f;
+
+  mvals = m_list[ind_second];
+  nvals = n_list[ind_second];
+  kvals = k_list[ind_second];
+
+  //Weighted phases for phase fusion weighted average
+  phi2_out = (t2 / 2.0f + mvals);
+  phi1_out = (t1 / 15.0f + kvals);
+  phi0_out = (t0 / 3.0f + nvals);  
+
+  *err_w2 = err_min_second;
+  *phase_second = (phi2_out + phi1_out + phi0_out) / 3.0f;
+
+  mvals = m_list[ind_third];
+  nvals = n_list[ind_third];
+  kvals = k_list[ind_third];
+
+  //Weighted phases for phase fusion weighted average
+  phi2_out = (t2 / 2.0f + mvals);
+  phi1_out = (t1 / 15.0f + kvals);
+  phi0_out = (t0 / 3.0f + nvals);
+
+  *err_w3 = err_min_third;
+  *phase_third = (phi2_out + phi1_out + phi0_out) / 3.0f;
+}
+
+
+void kernel processPixelStage2_phase3(global const float3 *a_in, global const float3 *b_in, global float *phase_1, global float *phase_2, global float *phase_3, global float *conf1, global float *conf2, global float *conf3)
+{
+  const uint i = get_global_id(0);
+
+  //read complex number real (a) and imaginary part (b)
+  float3 a = a_in[i];
+  float3 b = b_in[i];
+  
+  //calculate complex argument
+  float3 phase = atan2(b, a);
+  phase = select(phase, (float3)(0.0f), isnan(phase));
+  phase = select(phase, phase + 2.0f * M_PI_F, isless(phase, (float3)(0.0f)));
+
+  //calculate amplitude or the absolute value
+  float3 ir = sqrt(a * a + b * b) * AB_MULTIPLIER;
+  ir = select(ir, (float3)(0.0f), isnan(ir));
+  
+  float ir_sum = ir.x + ir.y + ir.z;
+
+  float phase_first = 0.0;
+  float phase_second = 0.0;
+  float phase_third = 0.0;
+  float J_1, J_2, J_3, unwrapping_likelihood1, unwrapping_likelihood2, unwrapping_likelihood3;
+
+  //scale with least common multiples of modulation frequencies
+  float3 t = phase / (2.0f * M_PI_F) * (float3)(3.0f, 15.0f, 2.0f);
+
+  float t0 = t.x;
+  float t1 = t.y;
+  float t2 = t.z;
+
+  //rank and extract three most likely phase hypothises
+  phaseUnWrapper3(t0, t1, t2, &phase_first, &phase_second, &phase_third, &J_1, &J_2, &J_3);
+
+  phase_1[i] = phase_first;
+  phase_2[i] = phase_second;
+  phase_3[i] = phase_third;
+
+  float phase_likelihood;
+  //check if near saturation
+  if(ir_sum < 0.4f * 65535.0f)
+  {
+    //calculate phase likelihood from amplitude
+    float var0, var1, var2;
+    calculatePhaseUnwrappingVar(ir, &var0, &var1, &var2);
+    phase_likelihood = exp(-(var0+var1+var2)/(2.0f * PHASE_CONFIDENCE_SCALE));
```

Comment:
Yes I guess we lose some precision in the exp operation. I am not sure about what you mean with the logsumexp?